kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/plugin/dsl/internal/VariantsImpl.kt

```diff
@@ -0,0 +1,171 @@
+        }
+
+        val variantConfig = customVariants.getOrPut(variantName) {
+            val variantConfig = objects.newInstance<KoverVariantCreateConfigImpl>()
+            variantConfig.deriveFrom(this)
+            variantConfig
+        }
+
+        block.execute(variantConfig)
+    }
+
+    override fun provided(variantName: String, block: Action<KoverVariantConfig>) {
+        if (variantName == TOTAL_VARIANT_NAME) {
+            throw KoverIllegalConfigException("The provided variant name cannot be empty.")
+        }
+
+        val variantConfig = providedVariants.getOrPut(variantName) {
+            val variantConfig = objects.newInstance<KoverVariantConfigImpl>()
+            variantConfig.deriveFrom(this)
+            variantConfig
+        }
+
+        block.execute(variantConfig)
+    }
+
+    override fun total(block: Action<KoverVariantConfig>) {
+        val variantConfig = providedVariants.getOrPut(TOTAL_VARIANT_NAME) {
+            objects.newInstance<KoverVariantConfigImpl>().also { newVariant ->
+                newVariant.deriveFrom(this)
+            }
+        }
+
+        block.execute(variantConfig)
+    }
+
+}
+
+internal abstract class KoverVariantConfigImpl @Inject constructor(objects: ObjectFactory) : KoverVariantConfig {
+    internal val classes: KoverVariantClasses = objects.newInstance()
+    internal val instrumentation: KoverVariantInstrumentation = objects.newInstance()
+    internal val testTasks: KoverVariantTestTasks = objects.newInstance()
+
+    override fun classes(block: Action<KoverVariantClasses>) {
+        block.execute(classes)
+    }
+
+    override fun instrumentation(block: Action<KoverVariantInstrumentation>) {
+        block.execute(instrumentation)
+    }
+
+    override fun testTasks(block: Action<KoverVariantTestTasks>) {
+        block.execute(testTasks)
+    }
+
+    internal fun deriveFrom(other: KoverVariantConfigImpl) {
+        classes.excludeJava.set(other.classes.excludeJava)
+        classes.excludedSourceSets.addAll(other.classes.excludedSourceSets)
+
+        instrumentation.disableAll.set(other.instrumentation.disableAll)
+        instrumentation.excludedClasses.addAll(other.instrumentation.excludedClasses)
+
+        testTasks.excluded.addAll(other.testTasks.excluded)
+    }
+}
+
+internal class MergingOptionality(val optional: Boolean, val withDependencies: Boolean)
+
+internal abstract class KoverVariantCreateConfigImpl @Inject constructor(private val objects: ObjectFactory) :
+    KoverVariantConfigImpl(objects), KoverVariantCreateConfig {
+    // variant name -> optionality
+    internal val variantsByName: MutableMap<String, MergingOptionality> = mutableMapOf()
+    internal val androidVariants: MutableList<Pair<KoverAndroidVariantBuilderImpl, MergingOptionality>> = mutableListOf()
+
+    override fun add(vararg variantNames: String, optional: Boolean) {
+        for (variantName in variantNames) {
+            addByName(variantName, optional, false)
```

Comment:
it was kind of a kind of trade-off, because `optional` is already `optional: Boolean = false`. Specifying two parameters with a default value doesn't look very good in DSL, and besides, it doesn't work well with Groovy, which is still in use.

Reply:
No, I meant here call site: `addByName(variantName, optional, withDependencies = false)`, so it will be more readable. I agree that adding two default parameters may be confusing.